src/pwm.rs

```diff
@@ -12,98 +12,136 @@ use core::time::Duration;
+///     let (pwm, [mut ch1, mut ch2]) = pwm.split(); 
+/// 
+///     ch1.try_set_duty(max_duty / 5).unwrap();
+///     ch2.try_set_duty(max_duty / 5).unwrap();
+///
 /// }
 ///
 /// # use core::convert::Infallible;
 /// # trait U32Ext { fn khz(self) -> Duration; }
 /// # impl U32Ext for u32 { fn khz(self) -> Duration { Duration::from_nanos( 1_000_000u64 / self as u64 ) } }
-/// # enum Channel { _1, _2 }
-/// # struct Pwm1;
-/// # impl hal::pwm::Pwm for Pwm1 {
+/// // Note: possibly you could genericise this over [PwmPinImpl; CHANNELS], which would be great for accessing channels internally
+/// // except, these are probably not all going to be the same type, so instead we're using marker traits to specify whether it has been split or not
+/// # struct Pwm1<S>(S);
+/// # struct Pwm1Pin;
+/// # struct HasPins;
+/// # impl Pwm1<HasPins> {
+/// #     pub fn split(self) -> (Pwm1<()>, [Pwm1Pin; 2]) {
+/// #       (Pwm1(()), [Pwm1Pin, Pwm1Pin])
+/// #     }
+/// # }
+/// # impl Pwm1<()> {
+/// #     pub fn join(self, pins: [Pwm1Pin; 2]) -> Pwm1<HasPins> {
+/// #       Pwm1(HasPins)
+/// #     }
+/// # }
+/// #
+/// # impl <S> hal::pwm::Pwm<u16> for Pwm1<S> {
 /// #     const CHANNELS: usize = 4;
+/// #     const MAX_DUTY: u16 = 1024;
 /// #     type Error = Infallible;
-/// #     type Duty = u16;
+/// #
 /// #     fn try_disable(&mut self, _: usize) -> Result<(), Self::Error> { unimplemented!() }
 /// #     fn try_enable(&mut self, _: usize) -> Result<(), Self::Error> { unimplemented!() }
-/// #     fn try_get_duty(&self, _: usize) -> Result<u16, Self::Error> { unimplemented!() }
-/// #     fn try_get_max_duty(&self) -> Result<u16, Self::Error> { Ok(0) }
-/// #     fn try_set_duty(&mut self, _: usize, _: u16) -> Result<(), Self::Error> { Ok(()) }
 /// #     fn try_get_period(&self) -> Result<Duration, Self::Error> { unimplemented!() }
 /// #     fn try_set_period<T>(&mut self, _: T) -> Result<(), Self::Error> where T: Into<Duration> { Ok(()) }
 /// # }
+/// #
+/// # impl hal::pwm::PwmDuty<u16> for Pwm1<HasPins> {
+/// #     fn try_get_duty(&self, _: usize) -> Result<u16, Self::Error> { unimplemented!() }
+/// #     fn try_set_duty<D>(&mut self, _: usize, _: D) -> Result<(), Self::Error> where D: Into<u16> { Ok(()) }
+/// # }
+/// # impl hal::pwm::PwmPin<u16> for Pwm1Pin {
+/// #     const MAX_DUTY: u16 = 1024;
+/// #     type Error = Infallible;
+/// #     fn try_disable(&mut self) -> Result<(), Self::Error> { unimplemented!() }
+/// #     fn try_enable(&mut self) -> Result<(), Self::Error> { unimplemented!() }
+/// #     fn try_get_duty(&self) -> Result<u16, Self::Error> { unimplemented!() }
+/// #     fn try_set_duty<D>(&mut self, _: D) -> Result<(), Self::Error> where D: Into<u16> { Ok(()) }
+/// # }
 /// ```
 // unproven reason: pre-singletons API. The `PwmPin` trait seems more useful because it models independent
 // PWM channels. Here a certain number of channels are multiplexed in a single implementer.
-pub trait Pwm {
+pub trait Pwm<Duty> {
```

Comment:
this is consistent with the manner we have implemented [spi](https://github.com/rust-embedded/embedded-hal/blob/master/src/blocking/spi.rs#L4) where `Duty` would reflect the word-size of the underlying timer and be implemented _once_ for each HAL. abstract consumers would require more complex bounds to support arbitrary duty types, but, this is somewhat inevitable for any abstraction and somewhat simplified by the use of a concrete type parameter vs. associated type w/ bounds ime.

Reply:
> this is consistent with the manner we have implemented [spi](https://github.com/rust-embedded/embedded-hal/blob/master/src/blocking/spi.rs#L4). `Duty` could be a primitive type reflecting the word-size of the underlying timer, implemented _once_ for each HAL (addressing #226).

If it's implemented once per HAL, then it should be an associated type. SPI (and serial) traits have this design, because it's reasonable to implement them multiple times per peripheral, as peripherals generally support multiple word sizes (I've done this in several HALs).

If there's one implementation-defined value of `Duty`, it should be an associated type. Associated types are less cumbersome to abstract over, in my experience, as they don't have to show up as a type parameter on the methods or impl blocks of the driver.